Toolchain part1: `./gen_mksurf_namelist.py`

negin513 opened the pull request:
### Description of changes
This Python script is part of the simplified toolchain for creating the surface dataset for global ctsm cases (#644 ).
(@slevisconsulting changed the word "certain" to "global" and added a comment about regional/single-point near the bottom of this post. I also changed the wording a little in the next paragraph and in the example a few lines down. I added a line under User Interface Changes.)

This script should be used as the first step of the new toolchain. 
It will automatically create namelist (control file) that is needed for creating the surface dataset (and requisite intermediate files) for running CTSM cases. 

For transient cases, it will also create a txt file that includes the landuse files for every year. 

Example: To run `gen_mksurf_namelist.py` for a transient case with resolution 4x5 for the years 2000-2005:

```
./gen_mksurf_namelist.py --start_year 2000 --end_year 2005 --res 4x5 --debug
```

### Specific notes

Contributors other than yourself, if any: 
@slevisconsulting 

CTSM Issues Fixed (include github issue #): #86 

Are answers expected to change (and if so in what way)?

Any User Interface Changes (namelist or namelist defaults changes)?
`./gen_mksurf_namelist.py` replaces aspects of `mksurfdata.pl`, which becomes obsolete with the new toolchain.
`./gen_mksurf_namelist.py` has the following user-interface:
```
usage: gen_mksurf_namelist.py [-h] [--sy START_YEAR] [--ey END_YEAR]
                              [--glc_nec GLC_NEC] [--rundir RUN_DIR]
                              [--ssp_rcp {hist,SSP1-2.6,SSP3-7.0,SSP5-3.4,SSP2-4.5,SSP1-1.9,SSP4-3.4,SSP4-6.0,SSP5-8.5}]
                              [--raw_dir INPUT_PATH] [-d] [-v] [--vic] [--glc]
                              [--hirespft] [--crop [CROP_FLAG]] [-f]
                              [-r {512x1024,360x720cru,128x256,64x128,48x96,94x192,0.23x0.31,0.47x0.63,0.9x1.25,1.9x2.5,2.5x3.33,4x5,10x15,0.125nldas2,5x5_amazon,1x1_camdenNJ,1x1_vancouverCAN,1x1_mexicocityMEX,1x1_asphaltjungleNJ,1x1_brazil,1x1_urbanc_alpha,1x1_numaIA,1x1_smallvilleIA,0.1x0.1,0.25x0.25,0.5x0.5,3x3min,5x5min,10x10min,0.33x0.33,0.125x0.125,ne4np4,ne16np4,ne30np4.pg2,ne30np4.pg3,ne30np4,ne60np4,ne120np4}]

|------------------------------------------------------------------|
|---------------------  Instructions  -----------------------------|
|------------------------------------------------------------------|
This Python script is part of the simplified toolchain for creating
the surface dataset for ctsm cases.
This script should be used as the first step of the new toolchain. 
It will automatically creating namelist (control  file) that is 
needed for creating surface dataset and other relevant files for
running CTSM cases. 
For transient cases, it will also create a txt file that includes the
landuse files for every year. 

-------------------------------------------------------------------
Instructions for running on Cheyenne/Casper:

load the following into your local environment:

    module load python
    ncar_pylib
-------------------------------------------------------------------
To see the available options:
    ./gen_mksurf_namelist.py --help

To run the script:
    ./gen_mksurf_namelist.py
 
To remove NPL from your environment on Cheyenne/Casper:
    deactivate
-------------------------------------------------------------------

optional arguments:
  -h, --help            show this help message and exit
  --sy START_YEAR, --start_year START_YEAR
                        Simulation start year. [default: 2000]
  --ey END_YEAR, --end_year END_YEAR
                        Simulation end year. [default: start_year]
  --glc_nec GLC_NEC     Number of glacier elevation classes to use. [default:
                        10]
  --rundir RUN_DIR      Directory to run in. [default:
                        /glade/scratch/negins/ctsm_toolchain/python/toolchain]
  --ssp_rcp {hist,SSP1-2.6,SSP3-7.0,SSP5-3.4,SSP2-4.5,SSP1-1.9,SSP4-3.4,SSP4-6.0,SSP5-8.5}
                        Shared Socioeconomic Pathway and Representative
                        Concentration Pathway Scenario name(s). [default:
                        hist]
  --raw_dir INPUT_PATH, --rawdata_dir INPUT_PATH
                        /path/of/root/of/input/data', [default:
                        /glade/p/cesm/cseg/inputdata/lnd/clm2/rawdata/]
  -d, --debug           Debug mode will print more information.
  -v, --verbose         Verbose mode will print more information.
  --vic                 Add the fields required for the VIC model. [default:
                        False]
  --glc                 Add the optional 3D glacier fields for verification of
                        the glacier model. [default: False]
  --hirespft            If you want to use the high-resolution pft dataset
                        rather than the default lower resolution dataset. (Low
                        resolution is at quarter-degree, high resolution at
                        3-minute) [Note: hires only available for 1850 and
                        2005.]
  --crop [CROP_FLAG]    Create datasets with the extensive list of prognostic
                        crop types. [default: False]
  -f, --fast            Toggle fast mode which does not user the large mapping
                        file
  -r {512x1024,360x720cru,128x256,64x128,48x96,94x192,0.23x0.31,0.47x0.63,0.9x1.25,1.9x2.5,2.5x3.33,4x5,10x15,0.125nldas2,5x5_amazon,1x1_camdenNJ,1x1_vancouverCAN,1x1_mexicocityMEX,1x1_asphaltjungleNJ,1x1_brazil,1x1_urbanc_alpha,1x1_numaIA,1x1_smallvilleIA,0.1x0.1,0.25x0.25,0.5x0.5,3x3min,5x5min,10x10min,0.33x0.33,0.125x0.125,ne4np4,ne16np4,ne30np4.pg2,ne30np4.pg3,ne30np4,ne60np4,ne120np4}, --res {512x1024,360x720cru,128x256,64x128,48x96,94x192,0.23x0.31,0.47x0.63,0.9x1.25,1.9x2.5,2.5x3.33,4x5,10x15,0.125nldas2,5x5_amazon,1x1_camdenNJ,1x1_vancouverCAN,1x1_mexicocityMEX,1x1_asphaltjungleNJ,1x1_brazil,1x1_urbanc_alpha,1x1_numaIA,1x1_smallvilleIA,0.1x0.1,0.25x0.25,0.5x0.5,3x3min,5x5min,10x10min,0.33x0.33,0.125x0.125,ne4np4,ne16np4,ne30np4.pg2,ne30np4.pg3,ne30np4,ne60np4,ne120np4}
                        Resolution is the supported resolution(s) to use for
                        files. [default: 4x5]
```

Single point and regional cases are now addressed separately: #1360 #1375 

Testing performed, if any:

negin513 commented:
@slevisconsulting Can you please review this code, before I ask others for a detailed code review?
Thanks for your help!

slevisconsulting commented:
Will do. Thank you for opening this PR @negin513. I will first go through your notes at the top of this page. I will mark my comments there with "@slevisconsulting".

slevisconsulting commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+# -[ ] hirespft data only for 2005?
+
+#QUESTIONS:
+# -[ ] Do we need --crop to accept y/n or just --crop is enough?
```
At the moment I'm leaning in favor of just --crop
because that's consistent with other options, such as --vic --hirespft --fast --glc
It may confuse users if only crop requires a y/n.

negin513 replied:
We discussed this for subset_data in a CSEG meeting. The consensus is to use --feature and --nofeature instead of --feature [y/n]. Here the crop flag is on by default so I introduced ---nocrop option.

negin513 replied:
The question for @ekluzek and others is that do we want crop to be on by default or off? 

I can easily change this to default no-crop which translate to having --crop option instead.

ekluzek replied:
The default should be crop on, since you can turn it off in CTSM. So almost all of the datasets we create will be with crop on. There will only be some exceptions to that where we'll create a dataset with crop off (one I can think of offhand are the VIC datasets). And a user might want to turn crop off explicitly.

slevisconsulting commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+        return parser
+
+
+def str2bool(v):
```
This function needed only if the group prefers
--crop y/n

slevisconsulting replied:
Also this function already exists elsewhere:
https://github.com/negin513/ctsm/blob/b854ad6b0b3e393cf4f1ab100edd3d085ef0cd14/tools/contrib/subset_data.py#L326-L349

So better to make a single copy of this function somewhere. I don't know where the correct location would be.

negin513 replied:
removed also the code is copied in utils.py.

slevisconsulting commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+def tag_describe ():
+    """
+    Function for converting different forms of
+    command line boolean strings to boolean value.
```
This is the description of a different function, so it needs to be replaced.

negin513 replied:
Thanks. Fixed!

slevisconsulting commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+
+class CtsmCase:
+    """
+    A class for encapsulate different ctsm cases.
```
```suggestion
    A class for encapsulating different ctsm cases.
```

negin513 replied:
Fixed!

slevisconsulting commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+    if (x < 850) or (x > 2105):
+        raise argparse.ArgumentTypeError(
+                "ERROR: Simulation start year should be between 850 and 2105.")
+    return x
```
Just a thought (this is nit-picky of me :-) )

Could the last two "_type" functions become one function where the range values and part of the error msg come in as arguments? Only if you feel like doing it.

negin513 replied:
I am not sure what you mean here. There are two functions here for (defining types) for command line arguments one is `glc_nec` and one is for `start_year`. I don't think we can combine them and use them in the parser and I don't see a point for doing that. This is how the code shows errors and warnings where glc_nec or year is not valid. I am marking this as resolved.

slevisconsulting reviewed (commented):
@negin513 nice work!

I am pausing my review at line 346 where you start describing the Methods. I will continue tomorrow.

slevisconsulting commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+        Build the namelist/control file for a ctsm class.  
+        """
+
+        self.landuse_filename()
```
Also, organizationally, I think it would make more sense to put the next two lines in `main` just before calling `build_nl`.

slevisconsulting commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+   '''
+   Figure out the dst mesh file for each res
+   '''
+   switcher = { 
```
Better to read the .xml file that contains the same info or is there a plan to remove this info from the .xml file?

slevisconsulting commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+        '1x1_asphaltjungleNJ','1x1_brazil,1x1_urbanc_alpha','1x1_numaIA,1x1_smallvilleIA','0.1x0.1','0.25x0.25','0.5x0.5',
+        '3x3min','5x5min','10x10min','0.33x0.33','0.125x0.125','ne4np4,ne16np4','ne30np4.pg2','ne30np4.pg3','ne30np4','ne60np4','ne120np4']
+        ,'ssp_rcp': ["hist","SSP1-2.6","SSP3-7.0","SSP5-3.4","SSP2-4.5","SSP1-1.9","SSP4-3.4","SSP4-6.0","SSP5-8.5"]
+        }
```
Or is there a plan to remove this info from the .xml file?
Ideally this info will exist in one location only.

slevisconsulting commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+                "start_year       = "+self.start_year.__str__() + "\n"
+                "end_year         = "+self.end_year.__str__() + "\n"
+                "mksrf_dyn_lu     = "+self.input_path+ "pftcftdynharv.0.25x0.25.LUH2.histsimyr1850-2015.c170629" + "\n"
+                "mksrf_fdynuse    = "+self.lu_fname + "\n"
```
This section needs some corrections. For example I ran `./mksurfdata.pl -years 1850-2000 -res 4x5` and:
- mksrf_dyn_lu doesn't exist
- mksrf_fdynuse  = 'landuse_timeseries_hist_78pfts_CMIP6_simyr1850-2015.txt'
- fdyndat        = 'landuse.timeseries_4x5_hist_78pfts_CMIP6_simyr1850-2015_c210703.nc'

negin513 replied:
I am not sure what you mean here. Do you mean there is an error with pl code?
When I run `./gen_mksurf_namelist -r 4x5 --sy 1850 --ey 2000`
The name list has:
mksrf_dyn_lu     = /glade/p/cesm/cseg/inputdata/lnd/clm2/rawdata/pftcftdynharv.0.25x0.25.LUH2.histsimyr1850-2015.c170629
mksrf_fdynuse    = landuse_timeseries_hist_78pfts_simyr1850-2000.txt

slevisconsulting commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+
+    #-- determine end_year if not given as an argument:
+    if not end_year:
+        end_year = start_year
```
You have something similar in `def __init__`
Does it not do the same thing?
Is it necessary to have both?

negin513 replied:
Correct. I removed it from __init__

slevisconsulting reviewed (commented):
I went through the whole thing. Again, thank you for your nice work, Negin! Let me know if you have questions about my review.

slevisconsulting commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+
+    def build_nl (self):
+        """
+        Build the namelist/control file for a ctsm class.  
```
```suggestion
        Build the namelist/control file for a ctsm case.
```
...though elsewhere I recommended removing the words "ctsm case" altogether.

negin513 replied:
I changed to the ctsm case here.

ekluzek commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+            #-- choose different files for years of 850-1850:
+            if (849 < yr < 1850):
+                lu_input_fname = os.path.join(self.input_path, 
+                            "pftcftdynharv.0.25x0.25.LUH2.histsimyr0850-1849.c171012",
```
I think it's important to have filenames in data files rather than hardcoded in the code. It could be the XML files as for namelist_defaults_ctsm_tools.xml, or in a different format, but should be in an outside data file. It is good to have it in fewer lines, but we have some development for more flexible data formats that allow this same type of thing in data files. Alper's mechanism for CESM or the new streams mechanism for NUOPC. One of those mechanisms could be used to shorten the number of lines required in the data file.

negin513 replied:
I agree Erik. I think maybe @slevisconsulting can implement this idea.

negin513 commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+This Python script is part of the simplified toolchain for creating
+the surface dataset for ctsm cases.
+This script should be used as the first step of the new toolchain. 
+It will automatically creating namelist (control  file) that is 
```
Fixed.

negin513 commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+the surface dataset for ctsm cases.
+This script should be used as the first step of the new toolchain. 
+It will automatically creating namelist (control  file) that is 
+needed for creating surface dataset and other relevant files for
```
Fixed!

negin513 commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+To run the script:
+    ./gen_mksurf_namelist.py
+ 
+To remove NPL from your environment on Cheyenne/Casper:
```
Fixed!

negin513 commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+
+    Raises:
+        Error if value of glc_start_year is not in the range 
+        of 850-2015. 
```
Thanks. Fixed!

negin513 commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+        x (str) : start_year string from command line args.
+
+    Raises:
+        Error if value of glc_start_year is not in the range 
```
Thanks. Fixed!

negin513 commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+        Build the land-use filename for a transient 
+        case.
+    create_landuse:
+        Create land-use txt file a transient case. 
```
Thanks. Fixed!

As you commanded, I changed the function names to:
```
build_landuse_filename
create_landuse_file
build_namelist_filename
create_namelist_file
```

negin513 commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+    -------
+    check_endyear:
+        Check if end_year is bigger than start year
+        in a ctsm case.
```
This is the class type definition for one ctsm case not multiple ctsm cases. These definitions are inside the type class so I included "a" ctsm case. 

For example the class person here is for one person and not multiple people. But we can instantiate it for multiple people:
```python
class Person:
"""
a class to define a person
"""
  def __init__(mysillyobject, name, age):
    mysillyobject.name = name
    mysillyobject.age = age

  def myfunc(abc):
    print("Hello my name is " + abc.name)
```
We can instantiate this class for multiple people:
```python
p1 = Person("John", 36)
p2 = Person("Jim", 56)
```
Similarly we can instantiate the class multiple times for multiple cases but the class definition is the definition for one.

negin513 commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+                    for item in self.__dict__))
+
+    def check_endyear (self):
+        if (int(self.end_year) < int(self.start_year)):
```
Fixed.

negin513 commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+        logging.debug(' run_type  = '+ self.run_type)
+
+    def check_num_pft (self):
+        #-- determine the num_pft
```
Fixed!

negin513 commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+        logging.debug(' crop_flag = '+ self.crop_flag.__str__()+ ' => num_pft ='+ self.num_pft)
+
+
+    def name_nl  (self):
```
fixed.

negin513 commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+    ctsm_case = CtsmCase(res, glc_nec, ssp_rcp, crop_flag, input_path,
+                         vic_flag, glc_flag, start_year, end_year)
+
+    ctsm_case.name_nl()
```
Thanks for catching this. although it did not have any impacts on the code, it was ran twice so I removed it.

negin513 commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+# -[ ] Possibly remove year --years and range options
+#      Currently comment them out.
+
+# -[ ] maybe a verbose option and removing debug
```
I used ctsm_logging now so I removed both explicit options which comes with ctsm_logging.

negin513 commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+        #            action="store",
+        #            dest="sim_range",
+        #            required=False,
+        #            choices = ['1850-2000','1850-2005','1850-2100'])
```
resolved.

negin513 commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+                    required=False,
+                    type = int)
+                    #type = end_year_type,
+                    #default="start_year")
```
commented out lines removed.

negin513 commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+    check_run_type:
+        Determine if a ctsm case is transient or
+        time-slice. 
+    def check_run_type:
```
fixed.

negin513 replied:
fixed.

negin513 commented on tools/toolchain/gen_mksurf_namelist.py:
```diff
@@ -0,0 +1,658 @@
+                            "_clm5_"+str(yr)+".c181217.nc")
+            else:
+                #--This error should never happen because
+                #  we have already checked years bound.
```
good catch! changed it 2106.

negin513 commented:
As of today, I think I addressed almost all the comments on this PR, except the comments related to a xml reader function for reading defaults or filenames. 

Also, I implemented all the feedback that I received from my other codes into this code, such as:

* Updating the flag arguments to --flag or --noflag and removing defaults. 
* Removing the progress bar when creating a land use file that showed the number of processed/remaining. 
* Moving the code into python/ctsm folder with top level skeleton code under tools/toolchain.
* Removing python 2 compatibility.
* Using CTSM/CIME logger. 
* CTSM pylint run is clean on this. 

slevisconsulting commented:
@negin513 thank you for the fast turnaround! This morning I emailed Scott Bachman and Isla Simpson to see if they also agree with our suggestion that I work on the PRs referenced in #644 (including this one and #1430 ) so that we may complete the pythonization of the mksurfdat toolchain. As soon as I have their go-ahead, I will start working on this one. Thanks again!

slevisconsulting reviewed (approved):
This PR is a successful first cut at generating a user-modifiable control file (namelist) for use with #1430 (mkmap_data.py) within the context of #644 (wrapper tool for generating fsurdat files). This PR (along with #1430) provides a great starting point in the overall development of a more user-friendly mksurfdat toolchain. Thank you @negin513 for your hard work at this!

slevisconsulting commented:
Clarification regarding my latest commit...
I eliminated what I assumed were pylint issues.
Now I wonder because `make all` gives me this error:
`Makefile:39: recipe for target 'lint' failed`

negin513 commented:
> Clarification regarding my latest commit... I eliminated what I assumed were pylint issues. Now I wonder because `make all` gives me this error: `Makefile:39: recipe for target 'lint' failed`

I previously cleaned up this PR so that pylint does not complain about anything. So the pylint should be clean on this. Just FYI so you don't spend a lot of time on it.

negin513 commented:
I am copying up my 2021-12-24 email here for the reference:

> Hello Sam, 
> PR 1 is ready for you to take over. You and Erik previously reviewed this and had some comments which I addressed. Thanks for reviewing the PR carefully. 
> I moved a step further and implemented all the feedback that I received from my other codes into this code, such as:
> Updating the flag arguments to --flag or --noflag and removing defaults. 
> Removing the progress bar when creating a land use file that showed the number of processed/remaining. 
> * Moving the code into python/ctsm folder with top level skeleton code under tools/toolchain.
> * Removing python 2 compatibility.
> * Using CTSM/CIME logger. 
> * CTSM pylint run is clean on this. 
> Overall this code is working (as it was previously) and can be merged anytime.  I think the next step for this code is: 
> * Review from Erik and Bill. Erik has already reviewed this before. 
> * One improvement to this code is a code/function for reading the xml files such as namelist_defaults_ctsm.xml to find available options. You and Erik made a similar suggestion and I agree it would be easier to handle them all in one place rather than keeping them in multiple places. I have some experience writing an xml parser code in 2015, so please let me know if you need help with writing this. 
> * Testing different options to make sure everything works as expected. I did some testing myself and it seems like it is working. 
> * I addressed all the comments on the github issue but the un-resolved comments are mostly related to the xml reader.
> 
> Please let me know how else I can help with this PR. 
> Negin

slevisconsulting commented:
Thank you, @negin513 , this is helpful.

slevisconsulting commented:
Hi @negin513 

1) Would you be able to look over what I wrote in the ChangeLog? Pls make changes if I misstated or forgot something. 
2) Then could you make a CTSM tag with the merge of this PR? Pls let me know if this is not feasible in the next day or two. (I realize that today is a holiday for NCAR employees.)

Thank you!

negin513 commented:
> Clarification regarding my latest commit... I eliminated what I assumed were pylint issues. Now I wonder because `make all` gives me this error: `Makefile:39: recipe for target 'lint' failed`
Hi @slevisconsulting, 
In the commit comment and doc/ChangeLog you mentioned the testing failed for this with `Makefile:39: recipe for target 'lint' failed`. 

When I test this, using `make` or `make lint` or `pylint -j 4 --rcfile=ctsm/.pylintrc ctsm` , it does not fail at all. I think maybe this error is caused by you not loading python 3.6+ when testing. I see similar issue when testing with python <3.6.

ekluzek commented:
Hmmm, thanks for looking into that @negin513. Python environments are tricky. We have an issue about testing different python versions, and this might point out the need to elevate that kind of work. We might also need to provide an conda environment that can be loaded for the python testing. When we end up in situations where it works differently for different people, it's something that we need to figure out and find a better solution for. This is a problem that comes up and it just indicates things need more work to remove those things that are dependent on user environment. It can be tricky to solve sometimes, but very worth it. So I'm actually glad that we ran into this, so that we can fix it.

And yes perfect answer @negin513 remove the line, and just document what your testing showed.

negin513 commented on python/ctsm/toolchain/ctsm_case.py:
```diff
@@ -266,122 +265,122 @@ def create_namelist_file(self):
 
         self.build_namelist_filename()
-        namelist_file = open(self.namelist_fname, "w")
+        with open(self.namelist_fname, "w", encoding='utf-8') as namelist_file:
```
@slevisconsulting I just wanted to say thanks for the update here from `open` to `with open`. I read about it a bit [online](https://stackoverflow.com/questions/34429519/python-difference-between-open-and-with-open) and it seems like using `with open` is the safer way to write a file compared to using `open`. I learnt this from your update, so just wanted to say thanks! 👍

slevisconsulting replied:
I found this from trying to eliminate pylint warnings :-)
